File: Projects/FlutterPerfLab/lib/core/managers/list_test_manager.dart
import 'dart:convert';
import 'package:flutter/foundation.dart'; // Required for compute()
import 'package:flutter/services.dart';
import '../../models/product_data.dart';

/// Top-level function for background isolate execution.
///
/// Offloads heavy JSON decoding from the Main UI Isolate. This prevents UI thread
/// blockage, minimizes CPU spikes, and isolates Garbage Collection churn
/// during the initial payload parsing.
List<ProductData> _parseJsonInBackground(String jsonString) {
  final List<dynamic> rawList = jsonDecode(jsonString) as List<dynamic>;

  return rawList
      .map((dynamic item) => ProductData.fromJson(item as Map<String, dynamic>))
      .toList(
        growable: false,
      ); // Pre-allocated fixed-size list for memory efficiency
}

/// A high-throughput, in-memory data provider engineered for virtualized UI
/// scrolling benchmarks (SliverList/ListView).
///
/// To ensure zero context-switching overhead on the UI thread, heavy I/O and
/// parsing operations are explicitly offloaded to a secondary isolate.
class ListTestManager {
  // Singleton instance
  static final ListTestManager shared = ListTestManager._internal();
  ListTestManager._internal();

  List<ProductData> _cachedData = [];
  int _cursor = 0;
  bool _isInitialized = false;

  /// Initializes the in-memory dataset asynchronously.
  Future<void> init() async {
    if (_isInitialized) return;

    try {
      // 1. Offload heavy File I/O via platform channels.
      final String jsonString = await rootBundle.loadString(
        'assets/mock_products.json',
      );

      // 2. Offload computational parsing to a background isolate.
      _cachedData = await compute(_parseJsonInBackground, jsonString);

      _isInitialized = true;
    } catch (e) {
      throw Exception(
        "CRITICAL: JSON decoding failed. Ensure mock_products.json is in assets. Error: $e",
      );
    }
  }

  /// Generates a deterministic page of mock products for the UI layer.
  ///
  /// List.generate with growable: false strictly pre-allocates contiguous memory
  /// for the page. This eliminates the O(n) overhead of dynamic array resizing
  /// during rapid scrolling.
  List<Product> fetchPage(int pageSize) {
    if (_cachedData.isEmpty) return [];

    // O(1) identity generation using the Event Loop's natural sequential execution.
    return List<Product>.generate(pageSize, (index) {
      final int uniqueId = _cursor;
      final int cacheIndex = uniqueId % _cachedData.length;

      // Cursor incremented synchronously.
      _cursor++;

      return Product(id: uniqueId, data: _cachedData[cacheIndex]);
    }, growable: false); // Enforces fixed-length memory allocation
  }

  /// Resets the pagination cursor to its initial state (0).
  /// Essential for reproducibility; guarantees that every benchmark session starts
  /// with the exact same data sequence.
  void resetCursor() {
    _cursor = 0;
  }
}

File: Projects/SwiftPerfLab/Core/Manager/ListTestManager.swift
import Foundation

/// A high-throughput data provider engineered for virtualized UI scrolling benchmarks.
actor ListTestManager {

    static let shared = ListTestManager()

    private var cachedData: [ProductData] = ListTestManager.parseJSON()

    private var cursor = 0

    private init() {}

    private static func parseJSON() -> [ProductData] {
        guard let url = Bundle.main.url(forResource: "mock_products", withExtension: "json") else {
            fatalError("❌ CRITICAL: mock_products.json not found in the main bundle.")
        }

        do {
            let data = try Data(contentsOf: url)
            let parsed = try JSONDecoder().decode([ProductData].self, from: data)
            print("✅ ListTestManager: Dataset cached. Item count: \(parsed.count)")
            return parsed
        } catch {
            fatalError("❌ CRITICAL: JSON decoding failed: \(error)")
        }
    }

    /// Marked as `async` to satisfy Swift 6 strict concurrency rules for actor state mutation.
    /// This guarantees thread safety and prevents the UI thread from deadlocking 
    /// during high-frequency pagination requests.
    func fetchPage(pageSize: Int) async -> [Product] {
        guard !cachedData.isEmpty else { return [] }

        var page: [Product] = []
        page.reserveCapacity(pageSize)

        for _ in 0..<pageSize {
            let item = cachedData[cursor % cachedData.count]
            page.append(Product(id: cursor, data: item))
            cursor += 1
        }
        return page
    }

    /// Safely resets the pagination cursor across the actor boundary,
    /// ensuring consistent state behavior for consecutive benchmark runs.
    func resetCursor() async {
        cursor = 0
    }
}